packages/auth/src/core/strategies/passkey.ts

```diff
@@ -0,0 +1,233 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Auth, User, UserCredential } from '../../model/public_types';
+
+import {
+  startPasskeyEnrollment,
+  StartPasskeyEnrollmentRequest,
+  StartPasskeyEnrollmentResponse,
+  finalizePasskeyEnrollment,
+  FinalizePasskeyEnrollmentRequest,
+  FinalizePasskeyEnrollmentResponse,
+  startPasskeySignIn,
+  StartPasskeySignInRequest,
+  StartPasskeySignInResponse,
+  finalizePasskeySignIn,
+  FinalizePasskeySignInRequest,
+  FinalizePasskeySignInResponse,
+  publicKeyCredentialToJSON
+} from '../../api/account_management/passkey';
+import { UserInternal } from '../../model/user';
+import { _castAuth } from '../auth/auth_impl';
+import { getModularInstance } from '@firebase/util';
+import { OperationType } from '../../model/enums';
+import { UserCredentialImpl } from '../user/user_credential_impl';
+import { signInAnonymously } from './anonymous';
+
+export async function signInWithPasskey(
+  auth: Auth,
+  name: string,
+  manualSignUp: boolean = false
+): Promise<UserCredential> {
+  const authInternal = _castAuth(auth);
+
+  // Start Passkey Sign in
+  const startSignInRequest: StartPasskeySignInRequest = {};
+  const startSignInResponse = await startPasskeySignIn(
+    authInternal,
+    startSignInRequest
+  );
+
+  const options = getPasskeyCredentialRequestOptions(startSignInResponse, name);
+
+  // Get the crendential
+  let credential;
+  try {
+    credential = (await navigator.credentials.get({
+      publicKey: options
+    })) as PublicKeyCredential;
+
+    const finalizeSignInRequest: FinalizePasskeySignInRequest = {
+      authenticatorAuthenticationResponse:
+        publicKeyCredentialToJSON(credential),
+      name,
+      displayName: name
+    };
+    const finalizeSignInResponse: FinalizePasskeySignInResponse =
+      await finalizePasskeySignIn(authInternal, finalizeSignInRequest);
+
+    const operationType = OperationType.SIGN_IN;
+    const userCredential = await UserCredentialImpl._fromIdTokenResponse(
+      authInternal,
+      operationType,
+      finalizeSignInResponse
+    );
+    await auth.updateCurrentUser(userCredential.user);
+    return userCredential;
+  } catch (error) {
+    if (
+      (error as Error).message.includes(
+        'The operation either timed out or was not allowed.'
```

Comment:
1. the credentials platform
2. it's thrown when there is no local passkey user or the user doesn't select any existing passkey. by 'current user' you mean Firebase current user?
3. it won't overwrite anything to the current user, will just sign in with the newly created user. I feel like we should leave the check of auth.currentUser to developers, instead of do the internal check in the sign in method.

Reply:
> by 'current user' you mean Firebase current user?

Yes

> it won't overwrite anything to the current user, will just sign in with the newly created user. 

signInAnonymously updates the current user - https://github.com/firebase/firebase-js-sdk/blob/ff1a6ec2c7a2a0dc79ea55eda6149c3dc5ca19ae/packages/auth/src/core/strategies/anonymous.ts#L56

> I feel like we should leave the check of auth.currentUser to developers, instead of do the internal check in the sign in method.

If they did not do this check and called signInWithPasskey, what will happen? I feel like there are 2 cases - 1) There is no firebase current user but a passkey is present. 2) No firebase user, no passkey.
What happens in case1? I understand that case 2 will throw the "operation either timed out..." error.

Reply:
> In case there is a current user and this error was thrown, we are now overwriting the uid with this new anonymously signed in user, right?

Let's say the current user is userA, the new anonymously signed in user is userB, so
1. Auth's current user will be updated to userB.
2. but none of userA's properties will be affected.

> 1) There is no firebase current user but a passkey is present. what will happen? 
User will be signed in with the passkey.